sphinx/util/docutils.py

```diff
@@ -113,3 +121,36 @@ def lookup_role(self, name, lang_module, lineno, reporter):
             return self.lookup_domain_element('role', name)
         except ElementLookupError:
             return self.role_func(name, lang_module, lineno, reporter)
+
+
+class WarningStream(object):
+    level_mapping = {
+        'DEBUG': logger.debug,
+        'INFO': logger.info,
+        'WARNING': logger.warning,
+        'ERROR': logger.error,
+        'SEVERE': logger.critical,
+    }
+
+    def write(self, text):
+        matched = report_re.search(text)
+        if not matched:
+            logger.warning(text.rstrip("\r\n"))
+        else:
+            location, type, level, message = matched.groups()
+            if type in self.level_mapping:
+                logger_method = self.level_mapping.get(type)
+                logger_method(message, location=location)
```

Comment:
Unfortunately, `logger.log()` expects integer levels.
But your idea seems good. I will implement this convert in our Adapter class.